src/url.rs

```diff
@@ -314,6 +358,120 @@ impl PyMultiHostUrl {
+            // check all of host / user / password / port empty
+            // build multi-host url
+            let mut multi_url = format!("{scheme}://");
+            for (index, single_host) in hosts.iter().enumerate() {
+                if single_host.is_empty() {
+                    return Err(PyValueError::new_err(
+                        "expected one of 'host', 'username', 'password' or 'port' to be set",
+                    ));
+                }
+                multi_url.push_str(&single_host.to_string());
+                if index != hosts.len() - 1 {
+                    multi_url.push(',');
+                };
+            }
+            multi_url
+        } else if let Some(host) = host {
+            let user_password = match (user, password) {
+                (Some(user), None) => format!("{user}@"),
+                (None, Some(password)) => format!(":{password}@"),
+                (Some(user), Some(password)) => format!("{user}:{password}@"),
+                (None, None) => String::new(),
+            };
+            let mut single_host_url = format!("{scheme}://{user_password}{host}");
+            if let Some(port) = port {
+                single_host_url.push(':');
+                single_host_url.push_str(port);
+            };
+            single_host_url
+        } else {
+            return Err(PyValueError::new_err("expected either `host` or `hosts` to be set"));
+        };
+
+        if let Some(path) = path {
+            url.push('/');
+            url.push_str(path);
+        }
+        if let Some(query) = query {
+            url.push('?');
+            url.push_str(query);
+        }
+        if let Some(fragment) = fragment {
+            url.push('#');
+            url.push_str(fragment);
+        }
+        cls.call1((url,))
+    }
+}
+
+pub struct MultiHostUrlHost {
+    username: Option<String>,
+    password: Option<String>,
+    host: Option<String>,
+    port: Option<String>,
+}
+
+impl MultiHostUrlHost {
+    fn is_empty(&self) -> bool {
+        self.host.is_none() && self.password.is_none() && self.host.is_none() && self.port.is_none()
+    }
+}
+
+impl FromPyObject<'_> for MultiHostUrlHost {
+    fn extract(ob: &'_ PyAny) -> PyResult<Self> {
+        let py = ob.py();
+        let dict = ob.downcast::<PyDict>()?;
+        Ok(MultiHostUrlHost {
+            username: dict.get_as(intern!(py, "username"))?,
+            password: dict.get_as(intern!(py, "password"))?,
+            host: dict.get_as(intern!(py, "host"))?,
+            port: dict.get_as(intern!(py, "port"))?,
+        })
+    }
+}
+
+impl fmt::Display for MultiHostUrlHost {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let mut url = String::new();
+        if self.username.is_some() && self.password.is_some() {
+            url.push_str(&format!(
+                "{}:{}",
+                self.username.as_ref().unwrap(),
+                self.password.as_ref().unwrap()
+            ));
```

Comment:
e.g. to write directly to the formatter:

```suggestion
            write!(
                f,
                "{}:{}",
                self.username.as_ref().unwrap(),
                self.password.as_ref().unwrap()
            )?;
```